ai21/clients/vertex/ai21_vertex_client.py

```diff
@@ -0,0 +1,192 @@
+
+
+class BaseAI21VertexClient:
+    def __init__(
+        self,
+        region: Optional[str] = None,
+        project_id: Optional[str] = None,
+        access_token: Optional[str] = None,
+        credentials: Optional[GCPCredentials] = None,
+    ):
+        if region is None:
+            region = DEFAULT_GCP_REGION
+
+        self._region = region
+        self._access_token = access_token
+        self._gcp_auth = GCPAuthorization(credentials=credentials, project_id=project_id)
+
+    def _get_base_url(self):
+        return VERTEX_BASE_URL_FORMAT.format(region=self._region)
+
+    def _get_access_token(self) -> str:
+        if self._access_token is not None:
+            return self._access_token
+
+        return self._gcp_auth.get_access_token()
+
+    def _build_path(
+        self,
+        project_id: str,
+        region: str,
+        model: str,
+        endpoint: str,
+    ):
+        return VERTEX_PATH_FORMAT.format(
+            project_id=project_id,
+            region=region,
+            model=model,
+            endpoint=endpoint,
+        )
+
+
+class AI21VertexClient(BaseAI21VertexClient, AI21HTTPClient):
+    def __init__(
+        self,
+        region: Optional[str] = None,
+        project_id: Optional[str] = None,
+        base_url: Optional[str] = None,
+        access_token: Optional[str] = None,
+        credentials: Optional[GCPCredentials] = None,
+        headers: Dict[str, str] | None = None,
+        timeout_sec: Optional[float] = None,
+        num_retries: Optional[int] = None,
+        http_client: Optional[httpx.Client] = None,
+    ):
+        BaseAI21VertexClient.__init__(
+            self,
+            region=region,
+            project_id=project_id,
+            access_token=access_token,
+            credentials=credentials,
+        )
+
+        if base_url is None:
+            base_url = self._get_base_url()
+
+        AI21HTTPClient.__init__(
+            self,
+            base_url=base_url,
+            timeout_sec=timeout_sec,
+            num_retries=num_retries,
+            headers=headers,
+            client=http_client,
+        )
+
+        self.chat = StudioChat(self)
+        # Override the chat.create method to match the completions endpoint,
+        # so it wouldn't get to the old J2 completion endpoint
+        self.chat.create = self.chat.completions.create
+
+    def _build_request(self, options: RequestOptions) -> httpx.Request:
+        options = self._prepare_options(options)
+
+        return super()._build_request(options)
+
+    def _prepare_options(self, options: RequestOptions) -> RequestOptions:
+        body = options.body
+
+        model = body.pop("model")
+        stream = body.pop("stream", False)
+        endpoint = "streamRawPredict" if stream else "rawPredict"
+        headers = self._prepare_headers()
+        path = self._build_path(
+            project_id=self._gcp_auth.project_id,
```

Comment:
`_project_id` should be an attribute of the client and not accessed via the auth object